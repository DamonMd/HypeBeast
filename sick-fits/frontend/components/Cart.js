import React from "react";
import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";
import SickButton from "./styles/SickButton";
import CartStyle from "./styles/CartStyles";
import Supreme from "./styles/Supreme";
import CloseButton from "./styles/CloseButton";

const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

const Cart = () => {
  return (
    <Mutation mutation={TOGGLE_CART_MUTATION}>
      {toggleCart => (
        <Query query={LOCAL_STATE_QUERY}>
          {({ data }) => (
            <CartStyle open={data.cartOpen} onClick={toggleCart}>
              <header>
                <CloseButton title="close">&times;</CloseButton>
                <Supreme>Your Cart</Supreme>
                <p> you have ....... items in your cart</p>
              </header>
              <footer>
                <p>$92.10</p>
                <SickButton>Checkout</SickButton>
              </footer>
            </CartStyle>
          )}
        </Query>
      )}
    </Mutation>
  );
};

export default Cart;
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION };
