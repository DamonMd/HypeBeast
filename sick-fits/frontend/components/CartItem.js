import React from "react";
import formatMoney from "../lib/formatMoney";
import styled from "styled-components";
import PropTypes from "prop-types";
import RemoveFromCart from "./RemoveFromCart";

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 1rem;
  }
  h3,
  p {
    margin: 0;
  }
`;

const CartItem = props => {
  // 1. check if item exist on the site still
  if (!props.cartItem.item) {
    return (
      <CartItemStyles>
        <p>This product has been removed from the site</p>
        <RemoveFromCart id={props.cartItem.id} />
      </CartItemStyles>
    );
  }
  const { image, price, title, id, description } = props.cartItem.item;

  return (
    <CartItemStyles>
      <img width="100" src={image} alt={title} />
      <div className="cart-item-details">
        <h3>{title}</h3>
        <p>
          {formatMoney(price * props.cartItem.quantity)} {" - "}{" "}
          <em>
            {props.cartItem.quantity} &times; {formatMoney(price)}
          </em>
        </p>
      </div>
      <RemoveFromCart id={props.cartItem.id} />
    </CartItemStyles>
  );
};

CartItem.propTypes = {
  cartItem: PropTypes.object.isRequired
};

export default CartItem;
