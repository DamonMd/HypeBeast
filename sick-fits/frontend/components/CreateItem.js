import React, { Component } from "react";
import Form from "./styles/Form";
import gql from "graphql-tag";
import Router from "next/router";
import { Mutation } from "react-apollo";
import Error from "./ErrorMessage";

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: "",
    description: "",
    price: 0,
    image: "",
    largeImage: ""
  };

  handleChange = e => {
    const { name, value, type } = e.target;
    const val = type === "number" ? parseFloat(value) : value;
    this.setState({ [name]: val });
  };

  uploadFile = async e => {
    const files = e.target.files;
    const data = new FormData();
    data.append("file", files[0]);
    //cloudinary config settings
    data.append("upload_preset", "hypebeast");
    const res = await fetch(
      "https://api.cloudinary.com/v1_1/dokwu1eus/image/upload",
      {
        method: "POST",
        body: data
      }
    );
    const file = await res.json();
    // const file = res;
    console.log(file);
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url
    });
  };

  render() {
    const { title, price, description, image } = this.state;
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
            onSubmit={async e => {
              e.preventDefault();
              //TODO: add in check to make sure cloudinary has returned a response
              //call mutation
              const response = await createItem();
              //re-route to single item page
              console.log(response);
              Router.push({
                pathname: "/item",
                query: { item: response.data.createItem.id }
              });
            }}
          >
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="title">Image</label>
              <input
                type="file"
                id="file"
                name="file"
                placeholder="Upload an image"
                onChange={this.uploadFile}
              />
              {this.state.image && (
                <img src={image} width="200" alt="Upload Preview" />
              )}
              <label htmlFor="title">Title</label>
              <input
                type="text"
                name="title"
                value={title}
                placeholder="Title"
                onChange={this.handleChange}
              />
              <label htmlFor="price">Price</label>
              <input
                type="number"
                name="price"
                value={price}
                onChange={this.handleChange}
              />
              <label htmlFor="description">Description</label>
              <textarea
                type="text"
                name="description"
                value={description}
                placeholder="Enter a description"
                onChange={this.handleChange}
              />
              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
