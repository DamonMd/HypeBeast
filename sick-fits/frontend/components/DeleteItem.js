import React, { Component } from "react";
import gql from "graphql-tag";
import { Mutation } from "react-apollo";
import { ALL_ITEMS_QUERY } from "./Items";

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

class DeleteItem extends Component {
  update = (cache, payload) => {
    //manually update the cache on client to match the server
    //read the cache for the items we want
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    console.log(data);
    console.log(payload);
    // filter deleted items out of the page
    data.items = data.items.filter(
      item => item.id !== payload.data.deleteItem.id
    );
    console.log("now the items areee....", data);
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data: data });
  };

  render() {
    const { id } = this.props;
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{ id: id }}
        update={this.update}
      >
        {(deleteItem, { error }) => (
          <button
            onClick={() => {
              if (confirm("you sure about deleting this?")) {
                deleteItem();
              }
            }}
          >
            {this.props.children}
          </button>
        )}
      </Mutation>
    );
  }
}

export default DeleteItem;
