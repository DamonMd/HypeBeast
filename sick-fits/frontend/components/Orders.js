import React, { Component } from "react";
import Error from "./ErrorMessage";
import gql from "graphql-tag";
import { Query } from "react-apollo";
import PropTypes from "prop-types";
import formatMoney from "../lib/formatMoney";
import Link from "next/link";
import styled from "styled-components";
import OrderItemStyles from "./styles/OrderItemStyles";

const ORDERS_QUERY = gql`
  query ORDERS_QUERY {
    orders {
      id
      total
      items {
        id
        price
        image
        quantity
        description
        title
      }
    }
  }
`;

const OrderUl = styled.ul`
  display: grid;
  grid-gap: 4rem;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`;

class Orders extends Component {
  render() {
    return (
      <div>
        <Query query={ORDERS_QUERY}>
          {({ data, loading, error }) => {
            if (error) return <Error error={error} />;
            if (loading) return <p>loading....</p>;
            console.log("data from order query", data);
            const orders = data.orders;
            return (
              <div>
                <h2>You have {orders.length} orders</h2>
                <OrderUl>
                  {orders.map(order => {
                    return (
                      <OrderItemStyles key={order.id}>
                        <Link
                          href={{
                            pathname: "/order",
                            query: { id: order.id }
                          }}
                        >
                          <a>
                            <div className="order-meta">
                              <p>
                                {order.items.reduce(
                                  (a, b) => a + b.quantity,
                                  0
                                )}{" "}
                                Items
                              </p>
                              <p>{formatMoney(order.total)}</p>
                            </div>
                            <div className="images">
                              {order.items.map(item => {
                                return (
                                  <img
                                    key={item.id}
                                    src={item.image}
                                    alt={item.description}
                                  />
                                );
                              })}
                            </div>
                          </a>
                        </Link>
                      </OrderItemStyles>
                    );
                  })}
                </OrderUl>
              </div>
            );
          }}
        </Query>
      </div>
    );
  }
}

export default Orders;
