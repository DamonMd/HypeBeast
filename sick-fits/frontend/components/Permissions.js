import React, { Component } from "react";
import { Query, MUTATION } from "react-apollo";
import gql from "graphql-tag";
import Error from "./ErrorMessage";
import Table from "./styles/Table";
import UserPermissions from "./UserPermissions";

const possiblePermissions = [
  "ADMIN",
  "USER",
  "ITEMCREATE",
  "ITEMUPDATE",
  "ITEMDELETE",
  "PERMISSIONUPDATE"
];

const USER_PERMISSION_QUERY = gql`
  query {
    users {
      name
      email
      id
      permissions
    }
  }
`;

const Permissions = props => {
  return (
    <Query query={USER_PERMISSION_QUERY}>
      {({ data, loading, error }) => {
        console.log(data);
        if (loading) return <p>loading...</p>;
        return (
          <div>
            <Error error={error} />
            <div>
              <h2>Manage Permissions</h2>
              <Table>
                <thead>
                  <tr>
                    <th>Name</th>
                    <th>Email</th>
                    {possiblePermissions.map((p, i) => (
                      <th key={i}>{p}</th>
                    ))}
                    <th>update</th>
                  </tr>
                </thead>
                <tbody>
                  {data.users.map(user => (
                    <UserPermissions
                      key={user.id}
                      possiblePermissions={possiblePermissions}
                      user={user}
                    />
                  ))}
                </tbody>
              </Table>
            </div>
          </div>
        );
      }}
    </Query>
  );
};

export default Permissions;
