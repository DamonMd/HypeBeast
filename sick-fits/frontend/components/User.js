import { Query } from "react-apollo";
import gql from "graphql-tag";
import PropTypes from "prop-types";
import SickButton from "./styles/SickButton";

const CURRENT_USER_QUERY = gql`
  query {
    me {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          price
          image
          title
          description
        }
      }
    }
  }
`;

const User = props => {
  return (
    <Query query={CURRENT_USER_QUERY}>
      {payload => props.children(payload)}
    </Query>
  );
};

const UserRights = props => {
  console.log("p", props);
  return (
    <tr>
      <td>{props.user.name}</td>
      <td>{props.user.email}</td>
      {props.permissions.map(p => (
        <td>
          <input type="checkbox" />
        </td>
      ))}
      <td>
        <SickButton>UPDATE</SickButton>
      </td>
      {/* {props.permissions.map(p => {
        return (
          <td>
            <label htmlFor={`${user.id}-permission-${p}`} />
            <input type="checkbox" />
          </td>
        );
      })} */}
    </tr>
  );
};

User.propTypes = {
  children: PropTypes.func.isRequired
};

export default User;
export { CURRENT_USER_QUERY };
export { UserRights };
