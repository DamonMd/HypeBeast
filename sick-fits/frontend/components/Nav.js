import React from "react";
import Link from "next/link";
import { Mutation } from "react-apollo";
import NavStyles from "./styles/NavStyles";
import User from "./User";
import Signout from "./Signout";
import { TOGGLE_CART_MUTATION } from "./Cart";
import CartCount from "./CartCount";

const Nav = () => {
  return (
    <User>
      {({ data: { me } }) => (
        <NavStyles data-test="nav">
          <Link href="/items">
            <a>shop</a>
          </Link>
          {me && (
            <>
              <Link href="/sell">
                <a>Sell</a>
              </Link>
              <Link href="/orders">
                <a>orders</a>
              </Link>
              <Link href="/me">
                <a>account</a>
              </Link>
              <Signout />
              <Mutation mutation={TOGGLE_CART_MUTATION}>
                {toggleCart => (
                  <button onClick={toggleCart}>
                    My Cart{" "}
                    <CartCount
                      count={me.cart.reduce((total, cart) => {
                        return total + cart.quantity;
                      }, 0)}
                    />
                  </button>
                )}
              </Mutation>
            </>
          )}
          {!me && (
            <Link href="/signup">
              <a>Sign In</a>
            </Link>
          )}
        </NavStyles>
      )}
    </User>
  );
};

export default Nav;
