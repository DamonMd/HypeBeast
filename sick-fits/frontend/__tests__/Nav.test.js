import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import Nav from "../components/Nav";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import { fakeUser, fakeCartItem } from "../lib/testUtils";
import { CURRENT_USER_QUERY } from "../components/User";

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: { me: null }
    }
  }
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: { me: fakeUser() }
    }
  }
];

const signedInMocksWithCartItems = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()]
        }
      }
    }
  }
];

describe("<Nav/>", () => {
  it("renders a sign in tab for logged out users", async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <Nav />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.text()).toContain("Sign In");
    const nav = wrapper.find('[data-test="nav"]');
    expect(toJSON(nav)).toMatchSnapshot();
  });

  it("renders full nav when signed in", async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <Nav />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const nav = wrapper.find('ul[data-test="nav"]');
    expect(nav.children().length).toBe(6);
    expect(nav.text()).toContain("Sign Out");
    // console.log(nav.debug());
  });

  //not really necessary since there is already a cart count test
  it("renders the amount of items in the cart", async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocksWithCartItems}>
        <Nav />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const nav = wrapper.find('ul[data-test="nav"]');
    const count = nav.find("CartCount");
    // console.log(count.debug());
    expect(toJSON(count)).toMatchSnapshot();
  });
});
