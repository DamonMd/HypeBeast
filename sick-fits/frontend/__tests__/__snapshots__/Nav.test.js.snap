// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders a sign in tab for logged out users 1`] = `
Array [
  <NavStyles
    data-test="nav"
  >
    <ul
      className="NavStyles-sc-11c0d2g-0 cAYLtA"
      data-test="nav"
    >
      <Link
        href="/items"
      >
        <a
          href="/items"
          onClick={[Function]}
        >
          shop
        </a>
      </Link>
      <Link
        href="/signup"
      >
        <a
          href="/signup"
          onClick={[Function]}
        >
          Sign In
        </a>
      </Link>
    </ul>
  </NavStyles>,
  <ul
    className="NavStyles-sc-11c0d2g-0 cAYLtA"
    data-test="nav"
  >
    <Link
      href="/items"
    >
      <a
        href="/items"
        onClick={[Function]}
      >
        shop
      </a>
    </Link>
    <Link
      href="/signup"
    >
      <a
        href="/signup"
        onClick={[Function]}
      >
        Sign In
      </a>
    </Link>
  </ul>,
]
`;

exports[`<Nav/> renders the amount of items in the cart 1`] = `
<CartCount
  count={9}
>
  <CartCount__AnimationStyles>
    <span
      className="CartCount__AnimationStyles-xxvp4g-0 gDkasO"
    >
      <TransitionGroup
        childFactory={[Function]}
        component="div"
      >
        <div>
          <CSSTransition
            className="count"
            classNames="count"
            in={true}
            key=".$9"
            onExited={[Function]}
            timeout={
              Object {
                "enter": 400,
                "exit": 400,
              }
            }
            unmountOnExit={true}
          >
            <Transition
              appear={false}
              className="count"
              enter={true}
              exit={true}
              in={true}
              mountOnEnter={false}
              onEnter={[Function]}
              onEntered={[Function]}
              onEntering={[Function]}
              onExit={[Function]}
              onExited={[Function]}
              onExiting={[Function]}
              timeout={
                Object {
                  "enter": 400,
                  "exit": 400,
                }
              }
              unmountOnExit={true}
            >
              <CartCount__Dot
                className="count"
              >
                <div
                  className="count CartCount__Dot-xxvp4g-1 ljNzVW"
                >
                  9
                </div>
              </CartCount__Dot>
            </Transition>
          </CSSTransition>
        </div>
      </TransitionGroup>
    </span>
  </CartCount__AnimationStyles>
</CartCount>
`;
